### DISCUSSION TITLE
Confused by "Machine User" docs

### DISCUSSION THREAD

**Discussion by sgpinkus** [community/community#24314] [Apps, API and Webhooks]
<p>I want to add a “machine user” because Travis suggest you do that to allow Travis to auto update the Github releases page post build:</p>
<p>&gt; “The suggested workaround is to create a <a href="https://developer.github.com/v3/guides/managing-deploy-keys/#machine-users" rel="nofollow noopener">machine user</a> — a dummy GitHub account that is granted write access on a per repository basis.”</p>
<p>The <a href="https://developer.github.com/v3/guides/managing-deploy-keys/#setup-3" rel="nofollow noopener">instructions to setup a machine user are</a>:</p>
<p>&gt; “Setup:<br>
&gt;   1. Run the ssh-keygen procedure on your server and attach the public key to the machine user account.<br>
&gt;   2. Give the machine user account access to the repositories you want to automate. You can do this by adding the account as a collaborator, as an outside collaborator, or to a team in an organization.”</p>
<p>I don’t get it. Is the “machine user account” merely the public 1/2 of an SSH key? If so I can’t figure out how to grant a user identified by a public SSH key access as a collaborator, as an outside collaborator, or to a team in an organization. Or is there a step missing: <em>Add an actual use and associate the SSH key with that user</em>? If so then why’s is called a machine user? That is just a user account.</p>

**Comment by ldnunes** [comment, marked as answer]
<p>Kinda necroing the topic, but anyway:</p>
<p>Yes, it’s still “just an user account”, but one that’s not accessed by humans, only scripts and/or programs. You will still have to create the account, which must have a valid email, and will also have to manage it’s password and SSH keys (in other words: safely store and use them). But the nice thing about it is that when using together with teams it gives you some nice flexibility for automating repository tasks, and it’s “totally cool” as GitHub puts it.</p>
<p>It’s a bit of work to set it up at the start, but after that it makes it easier to manage.</p>

**Comment by jeffwilcox** [comment]
<p>Also consider whether using a Deploy Key may make your scenario work… it’s essentially an SSH identity that is not associated with a user, so it does not require password management, etc., and is commonly used by web deployment platforms.</p>
<p>In our orgs, we’ve had to evolve our approach to using GitHub Apps and also deploy keys more often, since we now use GitHub Enterprise Cloud and also SAML-based single sign-on… having single sign-on for service accounts can be challenging in some organizations.</p>
<p>Docs - <a href="https://developer.github.com/v3/repos/keys/" rel="nofollow noopener">https://developer.github.com/v3/repos/keys/</a></p>

**Comment by nimishajn** [comment]
<p>I am also looking for a machine user for managing our organisation github repositories and we also have SSO enabled. Could you explain what would be the approach of creating deploy keys or github apps to start with? I could find several GitHub docs explaining what these “terms” mean but do not explain the implementation for SSO enabled Github organisations.<br>
Any URL would be helpful.<br>
Thanks</p>

**Comment by jeffwilcox** [comment]
<p>With the GitHub App, once installed on specific repos, or all repos on an org, it has the granular permissions you grant it. SSO won’t impact it.</p>

**Comment by amenocal** [comment]
<p><a class="mention" href="/jeffwilcox">@jeffwilcox</a> I’m interested in how you guys used GitHub Apps in your Enterprise w/SAML as “Machine Users”. Are there any references you could link me to in order to check them out?</p>
